python/tank/descriptor/bundle_cache_usage/writer_sqlite.py

```diff
@@ -0,0 +1,345 @@
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Methods relating to the Path cache, a central repository where metadata about
+all Tank items in the file system are kept.
+
+"""
+
+import sqlite3
+import os
+import time
+from datetime import datetime, timedelta
+
+from . import BundleCacheUsageLogger as log
+from .writer_base import BundleCacheUsageWriterBase
+
+
+class BundleCacheUsageSQLiteWriter(BundleCacheUsageWriterBase):
+    """
+    A local flat file SQLite-based database for tracking bundle cache accesses.
+    """
+
+    DB_MAIN_TABLE_NAME = "bundles"
+    DB_FILENAME = "bundle_usage.sqlite3"
+    DB_COL_ID = "id"
+    DB_COL_ID_INDEX = 0
+    DB_COL_PATH = "path"
+    DB_COL_PATH_INDEX = 1
+    DB_COL_ADD_DATE = "add_date"
+    DB_COL_ADD_DATE_INDEX = 2
+    DB_COL_LAST_ACCESS_DATE = "last_access_date"
+    DB_COL_LAST_ACCESS_DATE_INDEX = 3
+    DB_COL_ACCESS_COUNT = "access_count"
+    DB_COL_ACCESS_COUNT_INDEX = 4
+
+    def __init__(self, bundle_cache_root):
+        super(BundleCacheUsageSQLiteWriter, self).__init__(bundle_cache_root)
+        log.debug_db_inst("__init__")
+
+        self._bundle_cache_usage_db_filename = os.path.join(
+            self.bundle_cache_root,
+            BundleCacheUsageSQLiteWriter.DB_FILENAME
+        )
+
+        self._db_connection = None
+        self._connect()
+        self._create_main_table()
+
+        # this is to handle unicode properly - make sure that sqlite returns
+        # str objects for TEXT fields rather than unicode. Note that any unicode
+        # objects that are passed into the database will be automatically
+        # converted to UTF-8 strs, so this text_factory guarantees that any character
+        # representation will work for any language, as long as data is either input
+        # as UTF-8 (byte string) or unicode. And in the latter case, the returned data
+        # will always be unicode.
+        self._db_connection.text_factory = str
+
+    def _execute(self, sql_statement, tuple=None):
+        """
+        Connects the database if not already connected and execute the
+        specified SQL statement.
+
+        :param sql_statement:
+        :param tuple:
+        :return:
+        """
+        if tuple:
+            return self._get_cursor().execute(sql_statement, tuple)
+        else:
+            return self._get_cursor().execute(sql_statement)
+
+    def _create_main_table(self):
+        """
+        Create the database main table if it doesn't exists.
+
+        NOTE:
+        SQLite does not have a storage class set aside for storing dates and/or times.
+        Instead, the built-in Date And Time Functions of SQLite are capable of storing
+        dates and times as TEXT, REAL, or INTEGER values:
+
+        Reference:
+        https://sqlite.org/datatype3.html
+        """
+        sql_statement = \
+            """CREATE TABLE IF NOT EXISTS %s ( 
+                %s integer PRIMARY KEY,
```

Comment:
the preferred syntax is different - see https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/path_cache.py#L1339